src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,228 @@
+        private byte[] _buffer;
+        private IntPtr _directoryHandle;
+        private string _currentPath;
+        private bool _lastEntryFound;
+        private Queue<(IntPtr Handle, string Path)> _pending;
+        private GCHandle _pinnedBuffer;
+
+        /// <summary>
+        /// Encapsulates a find operation.
+        /// </summary>
+        /// <param name="directory">The directory to search in.</param>
+        public FindEnumerable(
+            string directory,
+            FindTransform<TResult> transform,
+            FindPredicate<TState> predicate,
+            TState state = default,
+            bool recursive = false)
+        {
+            _originalUserPath = directory;
+            _originalFullPath = Path.GetFullPath(directory);
+            _recursive = recursive;
+            _predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+            _transform = transform ?? throw new ArgumentNullException(nameof(transform));
+            _threadId = Environment.CurrentManagedThreadId;
+            _state = state;
+            Initialize();
+        }
+
+        private FindEnumerable(
+            string originalUserPath,
+            string originalFullPath,
+            FindTransform<TResult> transform,
+            FindPredicate<TState> predicate,
+            TState state,
+            bool recursive)
+        {
+            _originalUserPath = originalUserPath;
+            _originalFullPath = originalFullPath;
+            _predicate = predicate;
+            _transform = transform;
+            _state = state;
+            _recursive = recursive;
+            _threadId = Environment.CurrentManagedThreadId;
+            Initialize();
+        }
+
+        /// <summary>
+        /// Simple wrapper to allow creating a file handle for an existing directory.
+        /// </summary>
+        public static IntPtr CreateDirectoryHandle(string path)
+        {
+            IntPtr handle = Interop.Kernel32.CreateFile_IntPtr(
+                path,
+                Interop.Kernel32.FileOperations.FILE_LIST_DIRECTORY,
+                FileShare.ReadWrite | FileShare.Delete,
+                FileMode.Open,
+                Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle == IntPtr.Zero || handle == (IntPtr)(-1))
+            {
+                // Historically we throw directory not found rather than file not found
+                int error = Marshal.GetLastWin32Error();
+                if (error == Interop.Errors.ERROR_FILE_NOT_FOUND)
+                    error = Interop.Errors.ERROR_PATH_NOT_FOUND;
+
+                throw Win32Marshal.GetExceptionForWin32Error(error, path);
+            }
+
+            return handle;
+        }
+
+        public IEnumerator<TResult> GetEnumerator()
+        {
+            if (Interlocked.Exchange(ref _enumeratorCreated, 1) == 0 && _threadId == Environment.CurrentManagedThreadId)
```

Comment:
The `_threadId == Environment.CurrentManagedThreadId` check does not seem to provide any interesting guarantees here.

Reply:
Ok, was following pattern of Iterator.cs in the same project.

Reply:
> was following pattern of Iterator.cs in the same project.

It only uses the thread ID check.  This pattern is viable with either the interlocked or the thread ID check, but both aren't necessary.